File: R/xewma.arl.f.R
# Computation of EWMA ARLs (mean monitoring), returns function
xewma.arl.f <- function(l, c, mu, zr=0, sided="one", limits="fix", r=40) {
  if ( l<=0 | l>1 )             stop("l has to be between 0 and 2")
  if ( c<=0 )                   warning("usually, c has to be positive")
  if ( zr>c & sided=="one" )    stop("wrong reflexion border")
  if ( r<4 )                    stop("r is too small")
  ctyp <- pmatch(sided, c("one", "two")) - 1
  if ( is.na(ctyp) )            stop("invalid ewma type")
  ltyp <- -1 + pmatch(limits, c("fix", "vacl", "fir", "both", "Steiner", "stat", "fink", "limit", "fixW", "fixC"))
  if ( is.na(ltyp) )            stop("invalid limits type")
  if ( (sided=="one") & !(limits %in% c("fix", "vacl", "stat", "limit", "fixW")) )
                                stop("not supported for one-sided EWMA (not reasonable or not implemented yet")

  LENGTH <- 3*r
  zeug <- .C("xewma_arl_f", as.integer(ctyp), as.double(l), as.double(c), as.double(zr),
                            as.double(mu), as.integer(ltyp), as.integer(r),
                            ans=double(length=LENGTH), PACKAGE="spc")$ans
                            
  g <- zeug[1:r]                          
  w <- zeug[1:r + r]
  z <- zeug[1:r + 2*r]  
   
  arl <- Vectorize( function(x) 1 + sum( w * dnorm( ( z - (1-l)*x ) / l - mu)/l * g )  )
  
  arl   
}

File: src/xewma_arl_f.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <R.h>

#define ewma1 0
#define ewma2 1
#define fix 0
#define vacl 1
#define fir 2
#define both 3
#define steiner 4
#define stat 5
#define fink 6
#define elimit 7
#define waldmann 8
#define collocation 9

double xe2_iglarl_f(double l, double c, double mu, int N, double *g, double *w, double *z);

double *vector (long n);

void xewma_arl_f(int *ctyp, double *l, double *c, double *zr, double *mu, int *ltyp, int *r, double *zeug)
{ double *ARL, *w, *z, zahl=0.;
  int i;

 ARL = vector(*r);
 w   = vector(*r);
 z   = vector(*r); 
 
 for (i = 0; i < *r; i++) { w[i] = -1.; z[i] = 0.; ARL[i] = 0.; } /* init */
 
 if ( *ctyp==ewma2 && *ltyp==fix ) zahl =  xe2_iglarl_f(*l, *c, *mu, *r, ARL, w, z);
 
 for (i = 0; i < *r; i++) {
   zeug[i] = ARL[i];
   zeug[i + *r] = w[i];
   zeug[i + *r + *r] = z[i];
 }
   
 free(z);
 free(w);
 free(ARL);
 
 if ( fabs(zahl) > 1e-9 ) warning("trouble in xewma_arl [package spc]"); 
}  
